pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,223 @@
+	defer subscriber.Finish()
+
+	for {
+		select {
+		case update := <-subscriber.Updates():
+			if err := sendX509SVIDResponse(update, stream); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func sendX509SVIDResponse(update *cache.WorkloadUpdate, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer) (err error) {
+	resp, err := composeX509SVIDBySelectors(update)
+	if err != nil {
+		//log.WithError(err).Error("Could not serialize X.509 SVID response")
+		return status.Errorf(codes.Unavailable, "could not serialize response: %v", err)
+	}
+
+	if err := stream.Send(resp); err != nil {
+		//log.WithError(err).Error("Failed to send X.509 SVID response")
+		return err
+	}
+
+	return nil
+}
+
+func composeX509SVIDBySelectors(update *cache.WorkloadUpdate) (*delegatedidentityv1.SubscribeToX509SVIDsResponse, error) {
+	resp := new(delegatedidentityv1.SubscribeToX509SVIDsResponse)
+	resp.X509Svids = []*delegatedidentityv1.X509SVIDWithKey{}
+
+	for td, _ := range update.FederatedBundles {
+		resp.FederatesWith = append(resp.FederatesWith, td.IDString())
+	}
+
+	for _, identity := range update.Identities {
+		// TODO:  it doesn't work and always prints "false"
+		//fmt.Printf("processing id %s, %t\n", identity.Entry.SpiffeId, identity.Entry.Admin)
+		// Do not send admin nor downstream SVIDs to the caller
+		if identity.Entry.Admin || identity.Entry.Downstream {
+			continue
+		}
+
+		id, _ := idutil.IDProtoFromString(identity.Entry.SpiffeId)
+
+		keyData, err := x509.MarshalPKCS8PrivateKey(identity.PrivateKey)
+		if err != nil {
+			return nil, fmt.Errorf("marshal key for %v: %w", id, err)
+		}
+
+		svid := &delegatedidentityv1.X509SVIDWithKey{
+			X509Svid: &types.X509SVID{
+				Id:        id,
+				CertChain: x509util.RawCertsFromCertificates(identity.SVID),
+				// TODO: what if SVIDs 0 doesn't exist?
+				ExpiresAt: identity.SVID[0].NotAfter.Unix(),
+			},
+			X509SvidKey: keyData,
+		}
+		resp.X509Svids = append(resp.X509Svids, svid)
+	}
+	return resp, nil
+}
+
+func (s *Service) SubscribeToX509Bundles(req *delegatedidentityv1.SubscribeToX509BundlesRequest, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509BundlesServer) error {
+	ctx := stream.Context()
+
+	authorized, err := s.isCallerAuthorized(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to attest caller: %w", err)
+	}
+
+	if !authorized {
+		return status.Error(codes.PermissionDenied, "no authorized")
+	}
+
+	subscriber := s.manager.SubscribeToBundleChanges()
+
+	// send initial update....
+	for td, bundle := range subscriber.Value() {
+		resp := &delegatedidentityv1.SubscribeToX509BundlesResponse{
+			CaCertificates: map[string][]byte{td.IDString(): marshalBundle(bundle.RootCAs())},
+		}
+
+		if err := stream.Send(resp); err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-subscriber.Changes():
+			// TODO(Mauricio): cache logic to avoid sending updates for all bundles each time.
```

Comment:
I think this todo is now outdated based on the logic we've recently gained consensus on?

Reply:
Yeah, you are right. Just confirming, `subscriber.Next()` just return the bundle that were update, right? If so, I'll just remove this TODO. I didn't pay attention to this :(

Reply:
I believe it returns all bundles, not just those that were updated. Looking back at our conversation around the proto, I think that is the behavior we want? https://github.com/spiffe/spire-api-sdk/pull/8#discussion_r677862926

Reply:
you are right, I did a quick test and after deleting one bundle (from two bundle), the other one was returned... good, I'll delete this TODO :D